### ISSUE TITLE
Two clarifications on the progress bar implementation

### ISSUE THREAD

**Issue by DavisVaughan** [tidyverse/purrr#1047]
- I'd like to check and make sure that it was purposeful that `cli_progress_set()` uses `i` rather than `i + 1` here in the loop https://github.com/tidyverse/purrr/blob/8b2c3cf93c857065437372d28d2bd80799b4ee77/src/map.c#L40
- I'd also like to see if we should consider putting `cli_progress_set()` _after_ the call to `.f` (brought about by thinking about https://github.com/DavisVaughan/furrr/pull/248)

In particular, this example has two quirks:

```r
library(purrr)

x <- 1:5

map(x, .progress = TRUE, function(elt) {
  Sys.sleep(4)
})
```

- The progress bar doesn't show up at all until after the first iteration (i.e. about 4 seconds in). I don't know if this is a consequence of `show_after` or the fact that we are setting `i = 0` on the first iteration.
- The progress bar never actually hits 100%. On the last iteration we set `i = 4`, which corresponds to 80% which does show up, but then we do the last iteration and exit the loop. I assume that calling `cli_progress_done()` will mark it as 100%, and that does seem to be the case if you set `clear = FALSE`, but I still wonder if we should be using `i + 1` in the loop.

CC @gaborcsardi

**Comment by hadley** [comment]
That reasoning makes sense to me.

**Comment by gaborcsardi** [comment]
> I'd like to check and make sure that it was purposeful that cli_progress_set() uses i rather than i + 1 here in the loop

> I'd also like to see if we should consider putting cli_progress_set() after the call to .f (brought about by thinking about https://github.com/DavisVaughan/furrr/pull/248)

Yes, it is on purpose. We set it to `i` because we haven't done the `i`th step yet. I.e. we set it to `i` _after_ the `i`th step, just like you want.

We set it before the call to `.f`, so if the developer or the user sets `show_after = 0`, then it is shown before a potentially lengthy first iteration.

> The progress bar doesn't show up at all until after the first iteration (i.e. about 4 seconds in). I don't know if this is a consequence of show_after or the fact that we are setting i = 0 on the first iteration.

Yes, it is because of `show_after`.

> The progress bar never actually hits 100%. On the last iteration we set i = 4, which corresponds to 80% which does show up, but then we do the last iteration and exit the loop. I assume that calling cli_progress_done() will mark it as 100%, and that does seem to be the case if you set clear = FALSE, but I still wonder if we should be using i + 1 in the loop.

Yes, `cli_progress_done()` will mark it as 100%, and another `cli_progress_set()` call would be redundant.

**Comment by DavisVaughan** [comment]
Okay, that all makes sense to me. I agree that it is fine as is.

**Comment by gthlor** [comment]
does the progress bar works if there is a mutate in the map function?
